cli_tools/common/utils/logging/logger.go

```diff
@@ -15,50 +15,179 @@
+	// Reset when ReadOutputInfo info is called.
+	outputInfo *pb.OutputInfo
 
-// Log logs a message
-func (l stdoutLogger) Log(message string) {
-	fmt.Printf("%s %s\n", l.prefix, newLogEntry(message))
+	// timestampFormat is the format string used when writing the current time in user and debug messages.
+	timestampFormat string
+
+	// timeProvider is a function that returns the current time. Typically time.Now. Exposed for testing.
+	timeProvider func() time.Time
+}
+
+// WriteUser writes message to the underlying log.Logger, and then buffers the message
+// for inclusion in ReadOutputInfo().
+func (l *defaultToolLogger) WriteUser(message string) {
+	l.write(l.userPrefix, message)
+}
+
+// WriteDebug writes message to the underlying log.Logger, and then buffers the message
+// for inclusion in ReadOutputInfo().
+func (l *defaultToolLogger) WriteDebug(message string) {
+	l.write(l.debugPrefix, message)
 }
 
-// NewDefaultLogger creates a new logger that sends all messages to the default logger.
-func NewDefaultLogger() LoggerInterface {
-	return defaultLogger{}
+// WriteTrace buffers the message for inclusion in ReadOutputInfo().
+func (l *defaultToolLogger) WriteTrace(message string) {
+	l.trace = append(l.trace, message)
 }
 
-type defaultLogger struct{}
+// WriteMetric keeps non-nil fields from metric for inclusion in ReadOutputInfo().
+// Elements of list fields are appended to the underlying view.
+func (l *defaultToolLogger) WriteMetric(metric *pb.OutputInfo) {
+	proto.Merge(l.outputInfo, metric)
+}
+
+// Returns a view comprised of:
+//   - Calls to WriteMetric
+//   - All user, debug, and trace logs. User and debug logs are appended into a single
+//     member of OutputInfo.SerialLogs; each trace log is a separate member.
+// All buffers are cleared when this is called. In other words, a subsequent call to
+// ReadOutputInfo will return an empty object.
+func (l *defaultToolLogger) ReadOutputInfo() *pb.OutputInfo {
+	// Prepend stdout logs (user, debug) to the trace logs.
+	var combinedTrace []string
+	if l.userAndDebugBuffer.Len() > 0 {
+		combinedTrace = []string{l.userAndDebugBuffer.String()}
+		l.userAndDebugBuffer.Reset()
+	}
+	ret := &pb.OutputInfo{SerialOutputs: append(combinedTrace, l.trace...)}
+	l.trace = []string{}
+	proto.Merge(ret, l.outputInfo)
+	l.outputInfo = &pb.OutputInfo{}
+	return ret
+}
+
+// write a message to the underlying logger, and buffer it for inclusion in OutputInfo.SerialLogs.
+func (l *defaultToolLogger) write(prefix, message string) {
+	var logLineParts []string
+	if prefix != "" {
+		if strings.HasSuffix(prefix, ":") {
+			logLineParts = append(logLineParts, prefix)
+		} else {
+			logLineParts = append(logLineParts, prefix+":")
+		}
+	}
+	logLineParts = append(logLineParts, l.timeProvider().Format(l.timestampFormat), message)
+	logLine := strings.Join(logLineParts, " ")
+	l.output.Print(logLine)
+	l.userAndDebugBuffer.WriteString(logLine)
+	if !strings.HasSuffix(logLine, "\n") {
```

Comment:
what about l.output? DO you think a guaranteed "\n" is reasonable for output?

Reply:
The standard library guarantees a newline with `log.Logger.Print`; that said, I think it's clearer to put the newline prior to the write so that future readers aren't confused. Updated.